File: tabaco-y-ron-backend/app/api/v1/uploads.py
import secrets
from pathlib import Path

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status

from app.api.deps import require_admin_or_gestor
from app.core.config import settings

router = APIRouter(prefix="/uploads", tags=["uploads"])

ALLOWED_CONTENT_TYPES = {"image/jpeg", "image/png", "image/webp", "image/gif"}
EXTENSION_BY_TYPE = {
    "image/jpeg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/gif": ".gif",
}
MAX_BYTES = 5 * 1024 * 1024  # 5 MB


@router.post(
    "/imagen",
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_admin_or_gestor)],
)
async def upload_imagen(file: UploadFile = File(...)) -> dict[str, str]:
    if file.content_type not in ALLOWED_CONTENT_TYPES:
        raise HTTPException(
            status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
            f"Tipo no soportado. Permitidos: {sorted(ALLOWED_CONTENT_TYPES)}",
        )
    contents = await file.read()
    if len(contents) > MAX_BYTES:
        raise HTTPException(status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, "Imagen demasiado grande (>5MB)")

    upload_dir = Path(settings.UPLOAD_DIR)
    upload_dir.mkdir(parents=True, exist_ok=True)
    extension = EXTENSION_BY_TYPE[file.content_type]
    filename = f"{secrets.token_urlsafe(16)}{extension}"
    file_path = upload_dir / filename
    file_path.write_bytes(contents)

    public_url = f"{settings.PUBLIC_BASE_URL.rstrip('/')}/static/uploads/{filename}"
    return {"filename": filename, "url": public_url}

File: tabaco-y-ron-web/lib/api/uploads.ts
import { randomBytes } from "node:crypto";
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";

import { BadRequest, httpError } from "./errors";

export const UPLOAD_DIR = path.join(process.cwd(), "public", "static", "uploads");
export const PUBLIC_BASE = "/static/uploads";

const ALLOWED: Record<string, string> = {
  "image/jpeg": ".jpg",
  "image/png": ".png",
  "image/webp": ".webp",
  "image/gif": ".gif",
};

const MAX_BYTES = 5 * 1024 * 1024;

function tokenUrlSafe(bytes: number): string {
  return randomBytes(bytes).toString("base64url");
}

export async function saveUploadedImage(file: File): Promise<{ filename: string; url: string }> {
  const ext = ALLOWED[file.type];
  if (!ext) {
    throw httpError(
      415,
      `Tipo no soportado. Permitidos: ${Object.keys(ALLOWED).sort().join(", ")}`,
    );
  }
  const buf = Buffer.from(await file.arrayBuffer());
  if (buf.byteLength > MAX_BYTES) {
    throw httpError(413, "Imagen demasiado grande (>5MB)");
  }
  if (buf.byteLength === 0) throw BadRequest("Archivo vacío");

  await mkdir(UPLOAD_DIR, { recursive: true });
  const filename = `${tokenUrlSafe(16)}${ext}`;
  await writeFile(path.join(UPLOAD_DIR, filename), buf);

  return { filename, url: `${PUBLIC_BASE}/${filename}` };
}
